python/lsst/daf/butler/datastores/fileDatastoreClient.py

```diff
@@ -0,0 +1,58 @@
+__all__ = ("get_dataset_as_python_object", "FileDatastoreGetPayload")
+
+from typing import Any, Literal
+
+from lsst.daf.butler import DatasetRef, DimensionUniverse, Location, SerializedDatasetRef, StorageClass
+from lsst.daf.butler._compat import _BaseModelCompat
+from lsst.daf.butler.datastore.cache_manager import DatastoreDisabledCacheManager
+from lsst.daf.butler.datastore.stored_file_info import SerializedStoredFileInfo, StoredFileInfo
+from lsst.daf.butler.datastores.file_datastore.get import (
+    DatasetLocationInformation,
+    Mapping,
+    generate_datastore_get_information,
+    get_dataset_as_python_object_from_get_info,
+)
+
+
+class FileDatastoreGetPayloadFileInfo(_BaseModelCompat):
+    # TODO DM-41879: Allowing arbitrary URLs here is a severe security issue,
+    # since it allows the server to trick the client into fetching data from
+    # any file on its local filesystem or from remote storage using credentials
+    # laying around in the environment.  This should be restricted to only
+    # HTTP, but we don't yet have a means of mocking out HTTP gets in tests.
+    url: str
+    metadata: SerializedStoredFileInfo
+
+
+class FileDatastoreGetPayload(_BaseModelCompat):
+    datastore_type: Literal["file"]
+    file_info: list[FileDatastoreGetPayloadFileInfo]
+    dataset_ref: SerializedDatasetRef
+
+
+def get_dataset_as_python_object(
+    model: FileDatastoreGetPayload,
+    *,
+    universe: DimensionUniverse,
+    parameters: Mapping[str, Any] | None,
+    storageClass: StorageClass | str | None,
+) -> Any:
+    fileLocations: list[DatasetLocationInformation] = [
+        (Location(None, file_info.url), StoredFileInfo.from_simple(file_info.metadata))
+        for file_info in model.file_info
+    ]
+
+    ref = DatasetRef.from_simple(model.dataset_ref, universe=universe)
+    if storageClass is not None:
+        ref = ref.overrideStorageClass(storageClass)
+
+    datastore_file_info = generate_datastore_get_information(
+        fileLocations,
+        refStorageClass=ref.datasetType.storageClass,
+        refComponent=ref.datasetType.component(),
+        parameters=parameters,
+        dataId=ref.dataId,
+    )
+    return get_dataset_as_python_object_from_get_info(
+        datastore_file_info, ref=ref, parameters=parameters, cache_manager=DatastoreDisabledCacheManager()
```

Comment:
It's not using globals and is meant to be safe for use with multiple processes sharing a cache. I would not expect threads to be a problem -- it mostly scans the file system for its source of truth and when a file is being retrieved it puts it in a separate location to make sure another process doesn't delete the file whilst it's being read.

Reply:
Yeah so it seems that having a separate instance of DatastoreCacheManager per-thread is likely threadsafe, but sharing an instance across threads is not because of `self._cache_entries`.

Reply:
Yeah, it's a mutable object so that's always going to be a problem.